src/common/FastRational.h

```diff
@@ -640,32 +691,45 @@ inline void addition(FastRational& dst, const FastRational& a, const FastRationa
 inline void substraction(FastRational& dst, const FastRational& a, const FastRational& b) {
     if (a.wordPartValid() && b.wordPartValid()) {
         if (b.num == 0) {
-            dst.num = a.num;
-            dst.den = a.den;
-        }
-        else if (a.num == 0) {
-            if (b.num == WORD_MIN)
-                goto overflow;
-            dst.num = -b.num; // This overflows if b = -2147483648
+            dst = a;
+        } else if (a.num == 0) {
+            CHECK_WORD(dst.num, -lword(b.num));
             dst.den = b.den;
-        }
-        else if (b.den == 1) {
+        } else if (a.den == b.den and a.num == b.num) {
+            dst.num = 0;
+            dst.den = 1;
+        } else if (b.den == 1) {
+            // Maximum subtraction here is INT_MAX - (INT_MIN)*(UINT_MAX) = 2^63-1 which does not overflow lword
+            // Minimum subtraction here is INT_MIN - (INT_MAX)*(UINT_MAX) = -9223372028264841217 which does not underflow lword
+            // (a.num - b.num*a.den) / a.den is already canonicalized
             CHECK_WORD(dst.num, lword(a.num) - lword(b.num)*a.den);
             dst.den = a.den;
-        }
-        else if (a.den == 1) {
-            CHECK_WORD(dst.num, -lword(b.num) + lword(a.num)*b.den);
+        } else if (a.den == 1) {
+            // (a.num*b.den - b.den)/b.den is already canonicalized.
+            CHECK_WORD(dst.num, lword(a.num)*b.den - lword(b.num));
             dst.den = b.den;
-        }
-        else {
-            lword n = lword(a.num)*b.den - lword(b.num)*a.den;
```

Comment:
Was there a bug here before? Is it possible that this expression overflows `lword`?
Basically, I am asking if the `CHECK_SUB_OVERFLOWS_LWORD`s are really necessary.

Reply:
I added a regression test for the case `common == 1` which overflows.  I should check if I can find one for the case where `common > 1`.

Reply:
OK, I agree with your example (from the other discussion) that will overflow for `common == 1`, but I am pretty sure that it cannot overflow if `common > 1`:
At minimal `common == 2`, so `a.den / common < 2^31` and `(a.den / common) * b.num < 2^62`, similarly for `(b.den / common) * a/num < 2^62` and so their sum is `< 2^63` and it cannot overflow `lword`.
Is that correct?

Reply:
I am not sure.  I'll need to repeat the same exercise I did for `common == 1` to convince ourselves.  While doing it for the `common == 1` there seemed to be a lot of headroom, so I'm a little surprised if `common > 1` never over/underflows, but I might be wrong.

Reply:
I'm quite convinced now that you're right, the `common > 1` case does not require the `CHECK_SUB_OVERFLOWS_LWORD`.  I'm going to run the benchmarks on that version (likely f7ce1e1f) and comment here probably tomorrow.

Reply:
And for the record, if I remember correctly, this was the original bug reported in #197 which triggered this PR 😄